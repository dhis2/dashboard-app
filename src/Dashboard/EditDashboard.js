import React, { Fragment, Component } from 'react';
import { connect } from 'react-redux';
import { fromDashboards, fromSelected } from '../reducers';
import EditDashboardContent from './EditDashboardContent';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import EditBar from '../ControlBarContainer/EditBar';
import {
    acSetEditNewDashboard,
    acSetEditDashboard,
} from '../actions/editDashboard';

class EditDashboard extends Component {
    state = {
        initialized: false,
    };

    initEditDashboard = () => {
        if (this.props.mode === 'new') {
            this.setState({ initialized: true });
            this.props.setNewDashboard();
        } else {
            if (this.props.dashboard) {
                this.setState({ initialized: true });
                this.props.setEditDashboard(
                    this.props.dashboard,
                    this.props.items
                );
            }
        }
    };

    componentDidMount() {
        this.initEditDashboard();
    }

    componentDidUpdate() {
        if (!this.state.initialized) {
            this.initEditDashboard();
        }
    }

    render() {
        return (
            <Fragment>
                <EditBar />
                <DashboardVerticalOffset />
                <EditDashboardContent />
            </Fragment>
        );
    }
}

const mapStateToProps = state => {
    const selectedId = fromSelected.sGetSelectedId(state);
    const dashboard = selectedId
        ? fromDashboards.sGetById(state, selectedId)
        : null;

    const items = fromDashboards.sGetItems(state);

    return {
        dashboard,
        items,
    };
};

export default connect(mapStateToProps, {
    setNewDashboard: acSetEditNewDashboard,
    setEditDashboard: acSetEditDashboard,
})(EditDashboard);
