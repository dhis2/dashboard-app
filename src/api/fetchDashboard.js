import arrayClean from 'd2-utilizr/lib/arrayClean'
import { isViewMode } from '../modules/dashboardModes'
import { getCustomDashboards } from '../modules/getCustomDashboards'
import { withShape } from '../modules/gridUtil'
import {
    getIdNameFields,
    getListItemFields,
    getFavoritesFields,
} from './metadata'

const getDashboardItemsFields = () =>
    arrayClean([
        'id',
        'type',
        'shape',
        'x',
        'y',
        'width~rename(w)',
        'height~rename(h)',
        'messages',
        'text',
        'appKey',
        `${getListItemFields().join(',')}`,
        `${getFavoritesFields().join(',')}`,
    ])

const baseDashboardFields = arrayClean([
    'id',
    'displayName',
    'displayDescription',
    'favorite~rename(starred)',
    'access',
    'restrictFilters',
    'allowedFilters',
    'layout',
    'itemConfig',
    `dashboardItems[${getDashboardItemsFields().join(',')}]`,
])

export const viewDashboardQuery = {
    resource: 'dashboards',
    id: ({ id }) => id,
    params: {
        fields: baseDashboardFields,
    },
}

export const editDashboardQuery = {
    resource: 'dashboards',
    id: ({ id }) => id,
    params: {
        fields: arrayClean([
            ...baseDashboardFields,
            `user[${getIdNameFields({ rename: true }).join(',')}]`,
            'name',
            'description',
            'created',
            'favorite',
            'favorites',
            'lastUpdated',
            'href', // needed for d2-ui-translations-dialog
        ]),
    },
}

// Get more info about selected dashboard
export const apiFetchDashboard = async (
    dataEngine,
    id,
    { mode = null, forSave = false } = {}
) => {
    const query = isViewMode(mode) ? viewDashboardQuery : editDashboardQuery

    try {
        const dashboardData = await dataEngine.query(
            { dashboard: query },
            {
                variables: {
                    id,
                },
            }
        )

        const dashboard = dashboardData.dashboard

        if (!forSave) {
            return getCustomDashboards(
                Object.assign({}, dashboard, {
                    dashboardItems: withShape(dashboard.dashboardItems),
                })
            )[0]
        }
        return dashboard
    } catch (error) {
        console.log('Error: ', error)
    }
}
