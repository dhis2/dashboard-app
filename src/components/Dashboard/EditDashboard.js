import React, { Component } from 'react'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'

import { acSetEditDashboard } from '../../actions/editDashboard'
import { sGetSelectedId } from '../../reducers/selected'
import {
    sGetDashboardById,
    sGetDashboardItems,
    sDashboardsIsFetching,
} from '../../reducers/dashboards'
import DashboardVerticalOffset from './DashboardVerticalOffset'
import DashboardContent from './DashboardContent'
import EditBar from '../ControlBar/EditBar'
import NoContentMessage from '../../widgets/NoContentMessage'

export const Content = ({ updateAccess }) => {
    return updateAccess ? (
        <DashboardContent editMode={true} />
    ) : (
        <NoContentMessage text={i18n.t('No access')} />
    )
}

Content.propTypes = {
    updateAccess: PropTypes.bool,
}
export class EditDashboard extends Component {
    state = {
        initialized: false,
    }

    initEditDashboard = () => {
        if (this.props.dashboard) {
            this.setState({ initialized: true })
            this.props.setEditDashboard(this.props.dashboard, this.props.items)
        }
    }

    componentDidMount() {
        this.initEditDashboard()
    }

    componentDidUpdate() {
        if (!this.state.initialized) {
            this.initEditDashboard()
        }
    }

    getDashboardContent = () => {
        const contentNotReady =
            !this.props.dashboardsLoaded || this.props.id === null

        return (
            <div className="dashboard-wrapper">
                {contentNotReady ? null : (
                    <Content updateAccess={this.props.updateAccess} />
                )}
            </div>
        )
    }

    render() {
        return (
            <>
                <EditBar />
                <DashboardVerticalOffset editMode={true} />
                {this.getDashboardContent()}
            </>
        )
    }
}

EditDashboard.propTypes = {
    dashboard: PropTypes.object,
    dashboardsLoaded: PropTypes.bool,
    id: PropTypes.string,
    items: PropTypes.array,
    setEditDashboard: PropTypes.func,
    updateAccess: PropTypes.bool,
}

const mapStateToProps = state => {
    const id = sGetSelectedId(state)
    const dashboard = id ? sGetDashboardById(state, id) : null

    const updateAccess =
        dashboard && dashboard.access ? dashboard.access.update : false

    return {
        dashboard,
        id,
        updateAccess,
        items: sGetDashboardItems(state),
        dashboardsLoaded: !sDashboardsIsFetching(state),
    }
}

export default connect(mapStateToProps, {
    setEditDashboard: acSetEditDashboard,
})(EditDashboard)
