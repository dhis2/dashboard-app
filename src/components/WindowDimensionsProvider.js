import React, { createContext, useContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import debounce from 'lodash/debounce'

export const WindowDimensionsCtx = createContext(null)

const windowDims = () => ({
    height: window.innerHeight,
    width: window.innerWidth,
})

const WindowDimensionsProvider = ({ children }) => {
    const [dimensions, setDimensions] = useState(windowDims())

    useEffect(() => {
        const handleResize = debounce(() => {
            setDimensions(windowDims())
        }, 200)
        window.addEventListener('resize', handleResize)
        return () => {
            window.removeEventListener('resize', handleResize)
        }
    }, [])

    return (
        <WindowDimensionsCtx.Provider value={dimensions}>
            {children}
        </WindowDimensionsCtx.Provider>
    )
}

WindowDimensionsProvider.propTypes = {
    children: PropTypes.node.isRequired,
}

export default WindowDimensionsProvider

export const useWindowDimensions = () => {
    return useContext(WindowDimensionsCtx)
}
