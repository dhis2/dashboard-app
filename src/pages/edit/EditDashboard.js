import { useDataEngine } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { Layer, CenteredContent, CircularLoader } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { acSetEditDashboard } from '../../actions/editDashboard'
import { apiFetchDashboard } from '../../api/fetchDashboard'
import DashboardContainer from '../../components/DashboardContainer'
import NoContentMessage from '../../components/NoContentMessage'
import { useWindowDimensions } from '../../components/WindowDimensionsProvider'
import { EDIT } from '../../modules/dashboardModes'
import { setHeaderbarVisible } from '../../modules/setHeaderbarVisible'
import { isSmallScreen } from '../../modules/smallScreen'
import { sGetIsPrintPreviewView } from '../../reducers/editDashboard'
import LayoutPrintPreview from '../print/PrintLayoutDashboard'
import ActionsBar from './ActionsBar'
import ItemGrid from './ItemGrid'
import NotSupportedNotice from './NotSupportedNotice'
import classes from './styles/EditDashboard.module.css'
import TitleBar from './TitleBar'

const EditDashboard = props => {
    const dataEngine = useDataEngine()
    const { width } = useWindowDimensions()
    const [redirectUrl, setRedirectUrl] = useState(null)
    const [hasUpdateAccess, setHasUpdateAccess] = useState(null)
    const [isLoading, setIsLoading] = useState(true)

    useEffect(() => {
        const loadDashboard = async () => {
            try {
                const dashboard = await apiFetchDashboard(
                    dataEngine,
                    props.id,
                    { mode: EDIT }
                )
                props.setEditDashboard(dashboard)
                setHasUpdateAccess(dashboard.access?.update || false)
                setIsLoading(false)
            } catch (error) {
                setRedirectUrl(props.id ? `/${props.id}` : '/')
                setIsLoading(false)
            }
        }

        if (isSmallScreen(width)) {
            setRedirectUrl(props.id ? `/${props.id}` : '/')
            return
        }
        setHeaderbarVisible(true)

        loadDashboard()
    }, [props.id])

    if (redirectUrl) {
        return <Redirect to={redirectUrl} />
    }

    if (isLoading) {
        return (
            <Layer translucent>
                <CenteredContent>
                    <CircularLoader />
                </CenteredContent>
            </Layer>
        )
    }

    const renderGrid = () => {
        if (props.isPrintPreviewView) {
            return <LayoutPrintPreview fromEdit={true} />
        }
        return (
            <DashboardContainer>
                <TitleBar />
                <ItemGrid />
            </DashboardContainer>
        )
    }

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
                data-test="outer-scroll-container"
            >
                <ActionsBar />
                {hasUpdateAccess ? (
                    renderGrid()
                ) : (
                    <NoContentMessage text={i18n.t('No access')} />
                )}
            </div>
            <div className={classes.notice}>
                <NotSupportedNotice
                    message={i18n.t(
                        'Editing dashboards on small screens is not supported. Resize your screen to return to edit mode.'
                    )}
                />
            </div>
        </>
    )
}

EditDashboard.propTypes = {
    id: PropTypes.string,
    isPrintPreviewView: PropTypes.bool,
    setEditDashboard: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
    id: ownProps.match?.params?.dashboardId || null,
    isPrintPreviewView: sGetIsPrintPreviewView(state),
})

export default connect(mapStateToProps, {
    setEditDashboard: acSetEditDashboard,
})(EditDashboard)
