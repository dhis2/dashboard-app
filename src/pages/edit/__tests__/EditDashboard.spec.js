import { render, act } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Provider } from 'react-redux'
import { Router, Route } from 'react-router-dom'
import configureMockStore from 'redux-mock-store'
import { apiFetchDashboard } from '../../../api/fetchDashboard.js'
import WindowDimensionsProvider from '../../../components/WindowDimensionsProvider.js'
import EditDashboard from '../EditDashboard.js'

jest.mock('../../../api/fetchDashboard')

/* eslint-disable react/prop-types */
jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')

    return {
        __esModule: true,
        ...originalModule,
        CenteredContent: function Mock({ children }) {
            return <div className="ui-CenteredContent">{children}</div>
        },
    }
})

/* eslint-enable react/prop-types */

jest.mock(
    '../../../components/Notice.js',
    () =>
        function Mock() {
            return <div className="Notice" />
        }
)

jest.mock(
    '../ActionsBar',
    () =>
        function MockActionsBar() {
            return <div>ActionsBar</div>
        }
)

jest.mock(
    '../TitleBar',
    () =>
        function MockTitleBar() {
            return <div>TitleBar</div>
        }
)
jest.mock(
    '../ItemGrid',
    () =>
        function MockEditItemGrid() {
            return <div>ItemGrid</div>
        }
)

jest.mock(
    '../../print/PrintLayoutDashboard',
    () =>
        function MockLayoutPrintPreview() {
            return <div>LayoutPrintPreview</div>
        }
)

const mockStore = configureMockStore()

const renderWithRouterMatch = (
    ui,
    {
        route = '/',
        history = createMemoryHistory({ initialEntries: [route] }),
        store = {},
    } = {}
) => {
    return {
        ...render(
            <>
                <header />
                <Provider store={mockStore(store)}>
                    <WindowDimensionsProvider>
                        <Router history={history}>
                            <Route path={'edit/:id'} component={ui} />
                        </Router>
                    </WindowDimensionsProvider>
                </Provider>
            </>
        ),
    }
}

const dashboardId = 'rainbowdash'

test('EditDashboard renders dashboard', async () => {
    const promise = Promise.resolve()
    apiFetchDashboard.mockResolvedValue({
        id: dashboardId,
        access: { update: true },
    })
    const { container } = renderWithRouterMatch(EditDashboard, {
        route: `edit/${dashboardId}`,

        store: {
            editDashboard: {
                printPreviewView: false,
            },
        },
    })

    await act(() => promise)
    expect(container).toMatchSnapshot()
})

test('EditDashboard renders print preview', async () => {
    const promise = Promise.resolve()
    apiFetchDashboard.mockResolvedValue({
        id: dashboardId,
        access: { update: true },
    })
    const { container } = renderWithRouterMatch(EditDashboard, {
        route: `edit/${dashboardId}`,
        store: {
            editDashboard: {
                printPreviewView: true,
            },
        },
    })

    await act(() => promise)
    expect(container).toMatchSnapshot()
})

test('EditDashboard renders message when not enough access', async () => {
    const promise = Promise.resolve()
    apiFetchDashboard.mockResolvedValue({
        id: dashboardId,
        access: { update: false },
    })
    const { container } = renderWithRouterMatch(EditDashboard, {
        route: `edit/${dashboardId}`,
        store: {
            editDashboard: {
                printPreviewView: false,
            },
        },
    })

    await act(() => promise)
    expect(container).toMatchSnapshot()
})

test('EditDashboard does not render titlebar and grid if small screen', async () => {
    global.innerWidth = 480
    const promise = Promise.resolve()
    apiFetchDashboard.mockResolvedValue({
        id: dashboardId,
        access: { update: true },
    })
    const { container } = renderWithRouterMatch(EditDashboard, {
        route: `edit/${dashboardId}`,
        store: {
            editDashboard: {
                printPreviewView: false,
            },
        },
    })

    await act(() => promise)
    expect(container).toMatchSnapshot()
})
