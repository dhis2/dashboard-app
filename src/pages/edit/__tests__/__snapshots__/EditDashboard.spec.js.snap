// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDashboard does not render titlebar and grid if small screen 1`] = `
<div>
  <header />
</div>
`;

exports[`EditDashboard renders dashboard 1`] = `
<div>
  <header />
  <div
    class="container dashboard-scroll-container"
    data-test="outer-scroll-container"
  >
    <div>
      ActionsBar
    </div>
    <div
      class="container dashboard-scroll-container"
      data-test="inner-scroll-container"
    >
      <div>
        TitleBar
      </div>
      <div>
        ItemGrid
      </div>
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="Notice"
    />
  </div>
</div>
`;

exports[`EditDashboard renders message when not enough access 1`] = `
<div>
  <header />
  <div
    class="container dashboard-scroll-container"
    data-test="outer-scroll-container"
  >
    <div>
      ActionsBar
    </div>
    <div
      class="container"
    >
      No access
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="Notice"
    />
  </div>
</div>
`;

exports[`EditDashboard renders print preview 1`] = `
<div>
  <header />
  <div
    class="container dashboard-scroll-container"
    data-test="outer-scroll-container"
  >
    <div>
      ActionsBar
    </div>
    <div>
      LayoutPrintPreview
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="Notice"
    />
  </div>
</div>
`;
